File: backend/mcp_servers/gcp_manager_server.js
#!/usr/bin/env node

/**
 * GCP Manager MCP Server
 * 
 * Google Cloud APIs (Monitoring, Billing)를 통해 할당량 및 결제 상태를 모니터링합니다.
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
    CallToolRequestSchema,
    ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';
import { MetricServiceClient } from '@google-cloud/monitoring';
import { CloudBillingClient } from '@google-cloud/billing';
import dotenv from 'dotenv';

dotenv.config();

const monitoringClient = new MetricServiceClient();
const billingClient = new CloudBillingClient();

const server = new Server(
    {
        name: 'gcp-manager-mcp-server',
        version: '1.0.0',
    },
    {
        capabilities: {
            tools: {},
        },
    }
);

// 도구 목록 정의
server.setRequestHandler(ListToolsRequestSchema, async () => {
    return {
        tools: [
            {
                name: 'get_quota_usage',
                description: 'Get current quota usage for a specific GCP service (e.g., Generative Language API)',
                inputSchema: {
                    type: 'object',
                    properties: {
                        projectId: {
                            type: 'string',
                            description: 'GCP Project ID',
                        },
                        service: {
                            type: 'string',
                            description: 'Service name (default: generativelanguage.googleapis.com)',
                            default: 'generativelanguage.googleapis.com',
                        },
                    },
                    required: ['projectId'],
                },
            },
            {
                name: 'get_billing_info',
                description: 'Get billing account information and status for a project',
                inputSchema: {
                    type: 'object',
                    properties: {
                        projectId: {
                            type: 'string',
                            description: 'GCP Project ID',
                        },
                    },
                    required: ['projectId'],
                },
            },
        ],
    };
});

// 도구 실행 핸들러
server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;

    try {
        switch (name) {
            case 'get_quota_usage': {
                const { projectId, service = 'generativelanguage.googleapis.com' } = args;
                const name = monitoringClient.projectPath(projectId);

                // 최근 1시간 동안의 쿼터 사용량 조회 필터
                // 참고: 실제 Gemini API의 정확한 메트릭 이름은 서비스에 따라 다를 수 있음
                const filter = `metric.type = "serviceruntime.googleapis.com/quota/rate/net_usage" AND resource.labels.service = "${service}"`;

                const [timeSeries] = await monitoringClient.listTimeSeries({
                    name,
                    filter,
                    interval: {
                        startTime: {
                            seconds: Math.floor(Date.now() / 1000) - 3600,
                        },
                        endTime: {
                            seconds: Math.floor(Date.now() / 1000),
                        },
                    },
                });

                return {
                    content: [
                        {
                            type: 'text',
                            text: JSON.stringify({
                                service,
                                usageData: timeSeries,
                                message: timeSeries.length > 0 ? "Usage data retrieved successfully." : "No usage data found for the given interval. This might mean no requests were made recently."
                            }, null, 2),
                        },
                    ],
                };
            }

            case 'get_billing_info': {
                const { projectId } = args;
                const [billingInfo] = await billingClient.getProjectBillingInfo({
                    name: `projects/${projectId}`,
                });

                return {
                    content: [
                        {
                            type: 'text',
                            text: JSON.stringify(billingInfo, null, 2),
                        },
                    ],
                };
            }

            default:
                throw new Error(`Unknown tool: ${name}`);
        }
    } catch (error) {
        console.error(`Error executing tool ${name}:`, error);
        return {
            content: [
                {
                    type: 'text',
                    text: `Error: ${error.message}`,
                },
            ],
            isError: true,
        };
    }
});

async function main() {
    const transport = new StdioServerTransport();
    await server.connect(transport);
    console.error('GCP Manager MCP Server running on stdio');
}

main().catch((error) => {
    console.error('Server error:', error);
    process.exit(1);
});

File: backend/mcp_servers/gcp_manager_server.py
#!/usr/bin/env python3
"""
GCP Manager MCP Server

MCP 서버로 Google Cloud Platform 리소스를 관리합니다 (쿼터, 빌링).
"""
import json
import sys
from datetime import datetime, timedelta
from mcp.server.fastmcp import FastMCP
from google.cloud import monitoring_v3
from google.cloud import billing_v1
from dotenv import load_dotenv

from .lifecycle import setup_signal_handlers, log_startup

load_dotenv()

mcp = FastMCP("gcp-manager-mcp-server")


@mcp.tool()
def get_quota_usage(project_id: str, service: str = "compute.googleapis.com") -> str:
    """Get quota usage for a GCP service

    Args:
        project_id: GCP project ID
        service: GCP service name (default: compute.googleapis.com)

    Returns:
        JSON with service name, usage data, and optional message
    """
    client = monitoring_v3.MetricServiceClient()
    project_name = f"projects/{project_id}"

    # Query last hour of quota usage
    now = datetime.utcnow()
    interval = monitoring_v3.TimeInterval(
        {
            "end_time": {"seconds": int(now.timestamp())},
            "start_time": {"seconds": int((now - timedelta(hours=1)).timestamp())},
        }
    )

    # Build the metric filter
    metric_filter = f'metric.type="serviceruntime.googleapis.com/quota/rate/net_usage" AND resource.labels.service="{service}"'

    results = client.list_time_series(
        request={
            "name": project_name,
            "filter": metric_filter,
            "interval": interval,
            "view": monitoring_v3.ListTimeSeriesRequest.TimeSeriesView.FULL,
        }
    )

    usage_data = []
    for result in results:
        for point in result.points:
            usage_data.append({
                "timestamp": point.interval.end_time.isoformat(),
                "value": point.value.double_value or point.value.int64_value,
                "metric": dict(result.metric.labels),
            })

    return json.dumps({
        "service": service,
        "usageData": usage_data,
        "message": f"Retrieved quota usage for {service}" if usage_data else f"No quota usage data found for {service}"
    }, indent=2, default=str)


@mcp.tool()
def get_billing_info(billing_account_name: str) -> str:
    """Get billing information for a GCP billing account

    Args:
        billing_account_name: Billing account name (format: billingAccounts/XXXXXX-XXXXXX-XXXXXX)

    Returns:
        JSON with billing account information
    """
    client = billing_v1.CloudBillingClient()

    try:
        billing_account = client.get_billing_account(name=billing_account_name)

        result = {
            "name": billing_account.name,
            "displayName": billing_account.display_name,
            "open": billing_account.open,
            "masterBillingAccount": billing_account.master_billing_account if billing_account.master_billing_account else None,
        }

        return json.dumps(result, indent=2)
    except Exception as e:
        raise ValueError(f"Failed to retrieve billing info: {str(e)}")


if __name__ == "__main__":
    setup_signal_handlers()
    log_startup("GCP Manager MCP Server")
    mcp.run(transport="stdio")
